flume-ng-channels/flume-file-channel/src/test/java/org/apache/flume/channel/file/TestFileChannelErrorMetrics.java

```diff
@@ -0,0 +1,222 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.flume.channel.file;
+
+import junit.framework.Assert;
+import org.apache.commons.io.FileUtils;
+import org.apache.flume.ChannelException;
+import org.apache.flume.Transaction;
+import org.apache.flume.channel.file.instrumentation.FileChannelCounter;
+import org.apache.flume.event.EventBuilder;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+
+import static junit.framework.Assert.assertEquals;
+import static junit.framework.Assert.assertFalse;
+import static junit.framework.Assert.assertNotNull;
+import static junit.framework.Assert.assertTrue;
+
+public class TestFileChannelErrorMetrics extends TestFileChannelBase {
+
+  public TestFileChannelErrorMetrics() {
+    super(1);
+  }
+
+  /**
+   * This tests multiple successful and failed put and take operations
+   * and checks the values of the channel's counters.
+   */
+  @Test
+  public void testEventTakePutErrorCount() throws Exception {
+    final long usableSpaceRefreshInterval = 1;
+    FileChannel channel = Mockito.spy(createFileChannel());
+    Mockito.when(channel.createLogBuilder()).then(new Answer<Log.Builder>() {
+      @Override
+      public Log.Builder answer(InvocationOnMock invocation) throws Throwable {
+        Log.Builder ret = (Log.Builder) invocation.callRealMethod();
+        ret.setUsableSpaceRefreshInterval(usableSpaceRefreshInterval);
+        return ret;
+      }
+    });
+    channel.start();
+
+    FileChannelCounter channelCounter = channel.getChannelCounter();
+
+    Transaction tx = channel.getTransaction();
+    tx.begin();
+    channel.put(EventBuilder.withBody("test1".getBytes()));
+    channel.put(EventBuilder.withBody("test2".getBytes()));
+    tx.commit();
+    tx.close();
+    assertEquals(2, channelCounter.getEventPutAttemptCount());
+    assertEquals(2, channelCounter.getEventPutSuccessCount());
+    assertEquals(0, channelCounter.getEventPutErrorCount());
+
+    tx = channel.getTransaction();
+    tx.begin();
+    channel.take();
+    tx.commit();
+    tx.close();
+    assertEquals(1, channelCounter.getEventTakeAttemptCount());
+    assertEquals(1, channelCounter.getEventTakeSuccessCount());
+    assertEquals(0, channelCounter.getEventTakeErrorCount());
+
+    FileUtils.deleteDirectory(baseDir);
+    Thread.sleep(2 * usableSpaceRefreshInterval);
```

Comment:
I agree with @adenes on this. This cache is not refreshed by a background thread but essentially by the channel.put() call from the test itself.